eslint/eslint-patch/src/modern-module-resolution.ts

```diff
@@ -68,7 +68,7 @@ for (let currentModule = module; ; ) {
 
       // Make sure we actually resolved the module in our call path
       // and not some other spurious dependency.
-      if (path.join(eslintCandidateFolder, 'lib/cli-engine/cli-engine.js') === currentModule.filename) {
+      if (currentModule.filename.startsWith(eslintCandidateFolder)) {
```

Comment:
The only open question I have is whether `currentModule.filename` and `eslintCandidateFolder` are both alwahs fully resolved and have normalized/matching path separators, and that `eslintCandidateFolder` ends with a path separator.

Have you tested this both on POSIX and Windows? Can you confirm the terminating slash?

Reply:
Hi, thanks for the prompt review!

1. `currentModule.filename` and `eslintCandidateFolder` are both always fully resolved and have matching path separators on Windows and POSIX - I've confirmed this by hand
2. `eslintCandidateFolder` *never* ends in a path separator, I've confirmed this by hand and by reading the docs and considering how it's set up. My bad! I've adjusted this MR to add `path.sep` at the end, and have tested that on both Windows and POSIX
3. All that said, this doesn't work in Windows. But it doesn't make any better or worse - it just had no effect at all, at least in my testing. There was already a bug in Windows. If that bug were taken care of, this code would behave properly.

<details>
<summary>Details about what went wrong with Windows</summary>

Here's how I tested in Windows:

1. Rented an OVH VPS with a fresh Windows Server 2022 installation
2. Installed Node.js 20 and VS Code on it
3. Installed the eslint plugin for VS Code
4. Added the same repo I've been testing locally, which uses the latest version of eslint + a flat config file + eslint-config-next imported via [FlatCompat](https://www.npmjs.com/package/@eslint/eslintrc#Usage). Note that this repo works fine for me on macOS with the same set up and versions of everything, as long as I put in the patch here

I noticed a crucial difference between macOS and windows. This code works by walking up the stack of module imports like `module.parent.parent.parent...` and looking for a module with a `.filename` that's within an eslint installation. The issue is, no such file can be found. Instead, by the time eslint-patch is imported, entire stack of module imports has been truncated, and nothing before import-fresh appears:

<img width="504" alt="windows import stack does not include eslint" src="https://github.com/microsoft/rushstack/assets/1456118/4837fc57-cb9f-4bdc-8c8c-073fd1f7c18d">

For comparison, here's the import stack on my mac:

<img width="429" alt="macos import stack does include eslint" src="https://github.com/microsoft/rushstack/assets/1456118/e5712c65-0b0d-4656-a574-0875c3af8b57">

Note how it includes `node_modules/eslint/` (that bottom line actually is from the VS Code extension). I tried adding the same debugging to import-fresh, and that got me interesting results. Initially, it had the full stack. But later on, import-fresh got imported yet again, and when that happened and it logged its stack, it only showed itself and nothing before it:

<img width="479" alt="Screenshot 2023-08-07 at 10 17 08 PM" src="https://github.com/microsoft/rushstack/assets/1456118/18f3365b-95c3-4d33-a3f7-72fc902f34cb">

 I tried replacing import-fresh with a simple function that just calls require, thinking maybe it was the culprit - that didn't work.

Anyway, I've seen that this MR doesn't hurt anything in Windows either - see my very first 2 points at the very top of this comment.
</details>

The current code, as it is in the main branch and as it is in my branch, simply would never work on the Windows set up I tried. More substantial effort is required to get it working, and I'm personally not invested enough to get sort through it.

But I believe this MR does contain a valuable fix for anyone running into the same issue as me - where the existing code in eslint-patch is almost perfect, but just needs a slight adjustment to deal with a new change in eslint.

Reply:
Sounds good! Thanks for testing.

We'll have to investigate the issue on Windows separately. Thanks for doing some initial investigation.